core/drivers/clk/clk.c

```diff
@@ -0,0 +1,615 @@
+{
+	struct clk *clk;
+	unsigned int parent;
+	unsigned int alloc_size = sizeof(struct clk) +
+				  parent_count * sizeof(struct clk *);
+
+	clk = malloc(alloc_size);
+	if (!clk)
+		panic();
+
+	memset(clk, 0, alloc_size);
+
+	if (parent_count) {
+		clk->parents = (struct clk **)(clk + 1);
+		clk->num_parents = parent_count;
+		for (parent = 0; parent < parent_count; parent++)
+			clk->parents[parent] = parent_clks[parent];
+	}
+
+	clk->name = name;
+	clk->ops = ops;
+	clk->scmi_id = CLK_NO_SCMI_ID;
+
+	return clk;
+}
+
+void clk_free(struct clk *clk)
+{
+	free(clk);
+}
+
+static int clk_check(struct clk *clk)
+{
+	if (clk->ops->set_parent && !clk->ops->get_parent)
+		return 1;
+
+	if (clk->num_parents > 1 && !clk->ops->get_parent)
+		return 1;
+
+	if ((clk->flags & CLK_SET_RATE_PARENT) && !clk->num_parents)
+		return 1;
+
+	return 0;
+}
+
+static void clk_compute_rate(struct clk *clk)
+{
+	unsigned long parent_rate = 0;
+
+	clk->rate = 0;
+
+	if (clk->parent)
+		parent_rate = clk->parent->rate;
+
+	if (clk->ops->get_rate)
+		clk->rate = clk->ops->get_rate(clk, parent_rate);
+	else
+		clk->rate = parent_rate;
+}
+
+uint8_t clk_get_num_parents(struct clk *clk)
+{
+	return clk->num_parents;
+}
+
+struct clk *clk_get_parent_by_index(struct clk *clk, unsigned int pidx)
+{
+	if (pidx >= clk->num_parents)
+		return NULL;
+
+	return clk->parents[pidx];
+}
+
+static void clk_init_parent(struct clk *clk)
+{
+	uint8_t pidx;
+
+	if (clk->num_parents > 1) {
+		pidx = clk->ops->get_parent(clk);
+		if (pidx >= clk->num_parents) {
+			EMSG("get_parent returned an invalid value\n");
+			return;
+		}
+
+		clk->parent = clk->parents[pidx];
+	} else {
+		if (clk->num_parents)
+			clk->parent = clk->parents[0];
+	}
+}
+
+int clk_register(struct clk *clk)
+{
+	assert(!clk_check(clk));
+
+	SLIST_INSERT_HEAD(&clk_list, clk, link);
+
+	clk_init_parent(clk);
+	clk_compute_rate(clk);
+
+	if (clk->ops->is_enabled)
+		clk->enabled = clk->ops->is_enabled(clk);
+	else
+		clk->enabled = 0;
+
+	DMSG("Registered clock %s, freq %ld\n", clk->name, clk_get_rate(clk));
+
+	return 0;
+}
+
+unsigned long clk_get_rate(struct clk *clk)
+{
+	if (!clk)
+		return 0;
+
+	return clk->rate;
+}
+
+int clk_set_rate(struct clk *clk, unsigned long rate)
+{
+	int was_enabled;
+	unsigned long parent_rate = 0;
+
+	if (clk->flags & CLK_SET_RATE_GATE) {
+		was_enabled = clk->enabled;
+		if (was_enabled)
+			clk_disable(clk);
+	}
+
+	/*
+	 * No need to check for clk->parent since this is check at clock
+	 * registration
+	 */
+	if (clk->flags & CLK_SET_RATE_PARENT)
+		clk_set_rate(clk->parent, rate);
+
+	if (clk->ops->set_rate) {
+		if (clk->parent)
+			parent_rate = clk->parent->rate;
+
+		if (clk->ops->set_rate(clk, rate, parent_rate))
+			return -1;
+
+		return 0;
+	}
+
+	clk_compute_rate(clk);
+
+	if (clk->flags & CLK_SET_RATE_GATE) {
+		if (was_enabled)
+			clk_enable(clk);
+	}
+
+	return -1;
+}
+
+int clk_enable(struct clk *clk)
+{
+	struct clk *parent;
+
```

Comment:
i think this function and `clk_disable()` miss a reference counter, in case several consumers rely on a single clock.

Reply:
Indeed.